### PR TITLE
Define Kingdom internal measurement services for v2alpha public API.

### PR DESCRIPTION
**Pull request by SanjayVas**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)

### CODE REVIEW

**Comment by SanjayVas** [comment]
This is for #52.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbDopoEEJXSpWPQ3cF1:b-25b5ce) (no related file):*
I vaguely remember some discussion that this should be a single internal service, but I can't recall why.

---

**Review by wangyaopw — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @efoxepstein and @SanjayVas)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbDpN2c-RHDf-WohjCD:btk6n25) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I vaguely remember some discussion that this should be a single internal service, but I can't recall why.
</blockquote></details>

for simplicity I guess

---
___
*[src/main/proto/wfa/measurement/internal/kingdom/data_provider.proto, line 29 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDmn7k1gsQC2V4a2T-:-MbDmn7l5yXNBW_pRDI3:b-8jt8jr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/data_provider.proto#L29)):*
> ```protobuf
>   fixed64 external_public_key_certificate_id = 2;
> 
>   message Details {
> ```

do you want a `Details details = 3;` here
___
*[src/main/proto/wfa/measurement/internal/kingdom/measurement_consumer.proto, line 29 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDnKmH5nhqINJhhBqc:-MbDnKmH5nhqINJhhBqd:b-jwyeit) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurement_consumer.proto#L29)):*
> ```protobuf
>   fixed64 external_public_key_certificate_id = 2;
> 
>   message Details {
> ```

same here, a details field?
___
*[src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto, line 107 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDoEsKCvqFaFAa4Quf:-MbDoEsKCvqFaFAa4Qug:b-ssk1kd) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto#L107)):*
> ```protobuf
>   Filter filter = 1;
> 
>   int32 limit = 2;
> ```

add comment describing what this is for?
___
*[src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto, line 163 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDp9QHCWU5TuGvpyIy:-MbDp9QHCWU5TuGvpyIz:b3n24v2) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto#L163)):*
> ```protobuf
> }
> 
> message UpdatePublicKeyResponse {}
> ```

add `// Deliberately empty`?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: 0 of 9 files reviewed, 5 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/data_provider.proto, line 29 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDmn7k1gsQC2V4a2T-:-MbEAVM-8ubqqm3U_pdr:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/data_provider.proto#L29)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

do you want a `Details details = 3;` here
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/internal/kingdom/measurement_consumer.proto, line 29 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDnKmH5nhqINJhhBqc:-MbEAZy6B8K-1XGahCCE:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurement_consumer.proto#L29)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

same here, a details field?
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto, line 107 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDoEsKCvqFaFAa4Quf:-MbECwiO9BS6SPnjMlTD:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto#L107)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

add comment describing what this is for?
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto, line 163 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDp9QHCWU5TuGvpyIy:-MbEDIKHA87sSrXiNEaY:b-wcdj81) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto#L163)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

add `// Deliberately empty`?
</blockquote></details>

Will add if you feel strongly, but I tend to do the opposite and add a TODO if it's not deliberately empty. I don't want to imply that anything without a "Deliberately empty" comment is unintentional.

**Review by wangyaopw — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @efoxepstein and @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto, line 163 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDp9QHCWU5TuGvpyIy:-MbEFwg5BLlZJKjIZvRw:b-lardww) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/4716f2c6ad8b3bdc5e41f86a0cc3f52f84b19f15/src/main/proto/wfa/measurement/internal/kingdom/measurements_service.proto#L163)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Will add if you feel strongly, but I tend to do the opposite and add a TODO if it's not deliberately empty. I don't want to imply that anything without a "Deliberately empty" comment is unintentional.
</blockquote></details>

no strong opinion here. Just that some other empty response messages contain that comment. Both are fine for me.
___
*[src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt, line 26 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbEGlLH0DomUP1kme4e:-MbEGlLH0DomUP1kme4f:b-a2vi0l) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/ed148989b7797369adb5fde00585451a84ee3ffc/src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt#L26)):*
> ```Kotlin
> class KingdomSchemaTest : UsingSpannerEmulator(KINGDOM_SCHEMA) {
> 
>   @Test fun `database is created`() {}
> ```

is the original test broken?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

___
*[src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt, line 26 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbEGlLH0DomUP1kme4e:-MbEInOI8oS-XhbvT6wn:bgkqnnr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/ed148989b7797369adb5fde00585451a84ee3ffc/src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt#L26)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

is the original test broken?
</blockquote></details>

Yes. Inserting a MeasurementConsumer is slightly more complicated now, as you need to first insert a Certificate to satisfy the constraints.

We don't actually need the schema test to do anything more than validate the schema by attempting to create a database with it, which is what the base class does.

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 5 of 8 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 3 unresolved discussions (waiting on @SanjayVas and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbIEG2f2Zs3OAUr2eEq:b-fvy7tr) (no related file):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

for simplicity I guess
</blockquote></details>

We should figure this out. Are we expecting a separate single service for panel match and another for vid?

---

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 3 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbIGPmv23EFK-_gsMSb:bwd1s5t) (no related file):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

We should figure this out. Are we expecting a separate single service for panel match and another for vid?
</blockquote></details>

I tend to vote for smaller services if there's a good way to logically organize the methods. I know that there's some interactions that can cross top-level tables, e.g. UpdatePublicKey could end up writing to either MeasurementConsumers or a DataProviders.

If we go with splitting by the subject of each method, the same way we split up services by resource type in the public API, it results in the following:
1. DataProviders
2. EventGroups
3. MeasurementConsumers
4. Certificates
5. PublicKeys
6. Measurements
7. Requisitions
8. MeasurementLogEntries (not yet covered in this PR)

If that's too granular, we could organize by root table. We'd group together Certificates and PublicKeys since they interact with more than one root table and PublicKeys reference Certificates. This would result in:
1. DataProviders  (EventGroups)
2. MeasurementConsumers (Measurements, Requisitions, MeasurementLogEntries)
3. Certificates (PublicKeys)

---

File: src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt [hunk 1]
```diff
@@ -14,43 +14,14 @@
...
-      "SELECT MeasurementConsumerId, ExternalMeasurementConsumerId FROM MeasurementConsumers",
-      Struct.newBuilder()
-        .set("MeasurementConsumerId")
-        .to(3011)
-        .set("ExternalMeasurementConsumerId")
-        .to(1)
-        .build()
-    )
-  }
+  @Test fun `database is created`() {}
```
Comment (stevenwarejones) [inline src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt:26]: why is this removed?

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 4 unresolved discussions (waiting on @SanjayVas and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbMWF66-8gfpi_1OPi5:b-rqzvo1) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I tend to vote for smaller services if there's a good way to logically organize the methods. I know that there's some interactions that can cross top-level tables, e.g. UpdatePublicKey could end up writing to either MeasurementConsumers or a DataProviders.

If we go with splitting by the subject of each method, the same way we split up services by resource type in the public API, it results in the following:
1. DataProviders
2. EventGroups
3. MeasurementConsumers
4. Certificates
5. PublicKeys
6. Measurements
7. Requisitions
8. MeasurementLogEntries (not yet covered in this PR)

If that's too granular, we could organize by root table. We'd group together Certificates and PublicKeys since they interact with more than one root table and PublicKeys reference Certificates. This would result in:
1. DataProviders  (EventGroups)
2. MeasurementConsumers (Measurements, Requisitions, MeasurementLogEntries)
3. Certificates (PublicKeys)
</blockquote></details>

Root tables are kind of Spanner-specific so maybe that's not the best principle. But I guess we can have an internal logical resource hierarchy that coincidentally matches how Spanner is organized.

I don't feel strongly here, I just want us to have some decision to unblock panel match APIs.

---

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 4 unresolved discussions (waiting on @SanjayVas and @wangyaopw)

___
*[src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt, line 26 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDrq4o3mhNqEXdHQRq-r2-26:-MbMdcXR56wpzElJ01c0:baajiec) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/ed148989b7797369adb5fde00585451a84ee3ffc/src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/KingdomSchemaTest.kt#L26)):*
Reproducing the reply to @wangyaopw above:
> Inserting a MeasurementConsumer is slightly more complicated now, as you need to first insert a Certificate to satisfy the constraints.
>
>We don't actually need the schema test to do anything more than validate the schema by attempting to create a database with it, which is what the base class does.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 4 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbMeMUzCWU-CW-Xkpju:bu33dr5) (no related file):*
> Root tables are kind of Spanner-specific so maybe that's not the best principle. But I guess we can have an internal logical resource hierarchy that coincidentally matches how Spanner is organized.

Yeah, the IDs/parent-child relationships mean the concept extends beyond Spanner. I suppose I should have stated that the organization is by the root of each entity tree.

I started considering how to construct a base test class that's DB-agnostic. There will be cases that cross the services in either way of organizing mentioned above. For example, creating a MeasurementConsumer or a DataProvider requires first creating a Certificate. This lends itself to the idea of having a single service, but panel match has the same issue of needing to have a DataProvider in order to claim an exchange step. Therefore, we should either put everything for the same DB instance in a single service or be okay with test classes exercising more than just their primary subject service. (Or we extract helpers, but then we end up in a similar state as before with the DB wrapper interfaces.)

I'm now leaning towards the first option with more granular services, as this is closer to what we already have and matches how we organize things in the public API. For the DB-agnostic test base, some test classes may instantiate more than their primary subject service (e.g. MeasurementConsumers test depends on Certificates test, and PublicKeys test depends on DataProviders and MeasurementConsumers). Unless there are any objections, I'll move forward with this split.

---

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 4 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbMihwc2sSAlwt8IZzK:b4g182y) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> Root tables are kind of Spanner-specific so maybe that's not the best principle. But I guess we can have an internal logical resource hierarchy that coincidentally matches how Spanner is organized.

Yeah, the IDs/parent-child relationships mean the concept extends beyond Spanner. I suppose I should have stated that the organization is by the root of each entity tree.

I started considering how to construct a base test class that's DB-agnostic. There will be cases that cross the services in either way of organizing mentioned above. For example, creating a MeasurementConsumer or a DataProvider requires first creating a Certificate. This lends itself to the idea of having a single service, but panel match has the same issue of needing to have a DataProvider in order to claim an exchange step. Therefore, we should either put everything for the same DB instance in a single service or be okay with test classes exercising more than just their primary subject service. (Or we extract helpers, but then we end up in a similar state as before with the DB wrapper interfaces.)

I'm now leaning towards the first option with more granular services, as this is closer to what we already have and matches how we organize things in the public API. For the DB-agnostic test base, some test classes may instantiate more than their primary subject service (e.g. MeasurementConsumers test depends on Certificates test, and PublicKeys test depends on DataProviders and MeasurementConsumers). Unless there are any objections, I'll move forward with this split.
</blockquote></details>

Sorry, I have my dependencies backwards re: Certificates but you get the idea (the CreateCertificate method requires a Duchy, EDP, or MC to exist. What I was mixing up was that insertion of any of these requires insertion of the Certificate row first).

---

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: 7 of 23 files reviewed, 4 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbNHl5n8uUI3BsFlkAl:b-896fix) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Sorry, I have my dependencies backwards re: Certificates but you get the idea (the CreateCertificate method requires a Duchy, EDP, or MC to exist. What I was mixing up was that insertion of any of these requires insertion of the Certificate row first).
</blockquote></details>

Done.

---

**Review by efoxepstein — APPROVED** [review]
> Reviewed 16 of 16 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 4 unresolved discussions (waiting on @SanjayVas and @wangyaopw)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbDopoEEJXSpWPQ3cF0:-MbNWC8_6oILznx69hz7:bo5v7co) (no related file):*
> I'm now leaning towards the first option with more granular services

SGTM

---
___
*[src/main/proto/wfa/measurement/internal/kingdom/event_groups_service.proto, line 32 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbNVn8O-q3PQQq29HvF:-MbNVn8P-b2332QboSX1:bdo94vz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/d73e83d452c0b07f1f0ac1b05d8f07b442f0b3c3/src/main/proto/wfa/measurement/internal/kingdom/event_groups_service.proto#L32)):*
> ```protobuf
> 
>   // Streams EventGroups ordered by create time, ascending.
>   rpc StreamEventGroups(StreamEventGroupsRequest) returns (stream EventGroup);
> ```

While we're redoing all the APIs, do we want to keep the decision to use Streaming from internal APIs? I feel kind of mixed about it -- maybe paginated is simpler?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 4 unresolved discussions (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/event_groups_service.proto, line 32 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90#-MbNVn8O-q3PQQq29HvF:-MbNXBRY-55GGolSOCh0:bq43edb) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/d73e83d452c0b07f1f0ac1b05d8f07b442f0b3c3/src/main/proto/wfa/measurement/internal/kingdom/event_groups_service.proto#L32)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

While we're redoing all the APIs, do we want to keep the decision to use Streaming from internal APIs? I feel kind of mixed about it -- maybe paginated is simpler?
</blockquote></details>

Technically you would stream what would be a single page, using the `limit` and `created_after`/`updated_after` request fields. The main benefit of streaming is that you don't need to have the whole page in memory from the internal server end. For simplicity, we're not worrying about being optimal about response message size.

That said, if we're smart about max page size it may not be much of an issue in practice. I just figure if we're doing async DB reads, might as well stay async for as long as possible.

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 3 unresolved discussions (waiting on @oliver-amzn, @SanjayVas, @wangyaopw, and @zachcwc)

**Review by zachcwc — APPROVED** [review]
> Reviewed 4 of 8 files at r1, 3 of 4 files at r2, 16 of 16 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/90)* status: all files reviewed, 3 unresolved discussions (waiting on @oliver-amzn, @SanjayVas, and @wangyaopw)

**Pull request merged**